### PR TITLE
feat(api): automate first day [NTGL-303]

### PR DESCRIPTION
**Pull request by Rayahhhmed**
The new endpoint `/api/getstartdate` will go through the course data.json and get the start date of the first course and send a string in DD/MM/YYYY format. Sanity checked with T1 T2 and T3 data and it is fine to assume that this works! Finally, wanted to ask why can't Freerooms just change their code to move to 2023 data instead of staying back in 2022?

### CODE REVIEW

File: api/src/server.ts [hunk 1]
```diff
@@ -6,13 +6,14 @@ import * as Tracing from "@sentry/tracing";
...
 import { writeData } from "./write-data";
+import { getStartDate } from "./getStartDate";
 
 const app = express();
 const port = process.env.PORT || 3001;
 
 // Sentry configurations
 Sentry.init({
-    dsn: `${process.env.SENTRY_INGEST_URL}`,
+    dsn: `${process.env.INGEST_URL}`,
```
Comment (lhjt) [inline api/src/server.ts]: Why was this changed?

**Review by lhjt — CHANGES_REQUESTED** [review]

File: api/src/server.ts (hunk 1 above)
Comment (munjoonteo) [reply api/src/server.ts]: It wasn't meant to be changed I've asked Raiyan to fix it dw

Comment (Rayahhhmed) [reply api/src/server.ts]: Yep, it was mentioned to me earlier, thanks for the reminder nonetheless!

Comment (lhjt) [reply api/src/server.ts]: Ah sorry I wasn't aware; all good!

File: api/src/getStartDate.ts [hunk 2]
```diff
@@ -0,0 +1,144 @@
+import * as express from "express";
+import { data } from "./load-data";
+const FIRST_COURSE: Number = 0;
```
Comment (munjoonteo) [inline api/src/getStartDate.ts]: It's not necessary to declare the type of a variable when you are initialising it to something, unless it is an empty array or object.

File: api/src/getStartDate.ts [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+const termArray: string[] = ["Summer", "T1", "T2", "T3"];
+
+/**
+ * Some notes on the start date for the automation:
+ * To start of there are some assumptions which need to be documented:
```
Comment (munjoonteo) [inline api/src/getStartDate.ts]: "start  of" -> "start off"

File: api/src/getStartDate.ts [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+ * 1) The course that is selected as the first class of the term is the one
+ * that is used to get the term date, after
```
Comment (munjoonteo) [inline api/src/getStartDate.ts]: Indentation doesn't match point 2 below

File: api/src/getStartDate.ts [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+ * some sanity checks, it was concluded that it works!
+ *
+ * 2) ECON1101 is the class of interest for which we check if a
+ *    particular term data is out or not This is because it
```
Comment (munjoonteo) [inline api/src/getStartDate.ts]: Missing full stop before "This"

File: api/src/getStartDate.ts [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+ *    is offered every term. If this changes later, please
+ *    do change this code.
+ *
+ *
+ * -> There are two different start dates for the automation.
+ *    One for Notangles and the other for Freerooms.
+ *    Notangles will get the most recent term start date.
```
Comment (munjoonteo) [inline api/src/getStartDate.ts]: This explanation is kinda long-winded. What about "Notangles requires the start date of the latest term for which data is available. Freerooms requires the start date of the current term. This means that Freerooms will not break once new class data is released for the upcoming term.".

File: api/src/getStartDate.ts [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+/**
+ * Get the appropriate term date for Freerooms. It will get the current
+ * term date and not the new term date.
+ */
+const getStartDateFreerooms = (req: express.Request, res: express.Response) => {
+    try {
+        // Get the latest term and check the first class of the term
+        // The assumption is that the starting date is the start date of the first class
+        // of the term.
+        const currTerm = term(); // Get the current term that has the most recent data.
```
Comment (munjoonteo) [inline api/src/getStartDate.ts]: Please refrain from making line-by-line comments. In this case, sufficiently describing what the `term` function does in its own documentation will explain what this line of code does. Also may I suggest to clearly define what you mean by the "current" term and the "most recent" term because I think you mean to say the most recent term here. `currTerm` should probably be also called `mostRecentTerm` or something.

File: api/src/getStartDate.ts [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+        const termStartDate =
+            data.timetableData[currTerm][FIRST_COURSE]["classes"][0]["termDates"]["start"];
+        const currDate = new Date(); // We will check if the current date is after the
```
Comment (munjoonteo) [inline api/src/getStartDate.ts]: This comment is unnecessary too. It could go in the docs above the function definition to give a general overview of what this function actually does.

File: api/src/getStartDate.ts [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+        // term start date so Freerooms does not suffer.
+        // Regex expression to match the timeDelta format.
+        let regexp = /(\d{2})\/(\d{2})\/(\d{4})/;
```
Comment (munjoonteo) [inline api/src/getStartDate.ts]: The comment above this line is unnecessary - the variable could just be called `timeRegex` or something. Also, this variable can be declared with `const` since it will not be modified. Go through the rest of your code and see if you can spot anything which can be changed to `const` instead of `let` too.

File: api/src/getStartDate.ts [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+        let day, month, year;
+        let matched = termStartDate.match(regexp);
+        if (matched != null) {
+            year = matched[3];
```
Comment (munjoonteo) [inline api/src/getStartDate.ts]: You can use array destructuring here. Also what's in `matched[0]` if not the day?

File: api/src/getStartDate.ts [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+            month = matched[2];
+            day = matched[1];
+        } else {
+            res.send("Error getting the start date");
+        }
+        const termStartDateObj = new Date(year, month, day);
+        // Getting the difference in time so Freerooms gets the most recent term start date when
+        // it starts.
+        const timeDelta = currDate.valueOf() - termStartDateObj.valueOf();
```
Comment (munjoonteo) [inline api/src/getStartDate.ts]: This variable could be called `isCurrentTerm` and be a boolean `currDate.valueOf() - termStartDateObj.valueOf() >= 0`. Then the comment is unnecessary and the following if condition will make more sense intuitively.

File: api/src/getStartDate.ts [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+
+        if (timeDelta >= 0) {
+            // If the data is more than a day old, we will return the term start date that is for the
```
Comment (munjoonteo) [inline api/src/getStartDate.ts]: This comment could go outside the if/else block to explain what you are trying to do.

File: api/src/getStartDate.ts [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+            // new term, else we get the date for the old term.
+            res.send(termStartDate);
+        } else {
+            let dummyTerm = termArray.findIndex((term) => term == currTerm) - 1;
```
Comment (munjoonteo) [inline api/src/getStartDate.ts]: Always use `===`

File: api/src/getStartDate.ts [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+ * term date and not the new term date.
+ */
+const getStartDateFreerooms = (req: express.Request, res: express.Response) => {
+    try {
+        // Get the latest term and check the first class of the term
+        // The assumption is that the starting date is the start date of the first class
+        // of the term.
+        const currTerm = term(); // Get the current term that has the most recent data.
+        const termStartDate =
+            data.timetableData[currTerm][FIRST_COURSE]["classes"][0]["termDates"]["start"];
```
Comment (munjoonteo) [inline api/src/getStartDate.ts]: Can accessing the start date be a function lol

File: api/src/getStartDate.ts [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+        const timeDelta = currDate.valueOf() - termStartDateObj.valueOf();
+
+        if (timeDelta >= 0) {
+            // If the data is more than a day old, we will return the term start date that is for the
+            // new term, else we get the date for the old term.
+            res.send(termStartDate);
+        } else {
+            let dummyTerm = termArray.findIndex((term) => term == currTerm) - 1;
+            // Checking if the term is the first term, we can have case when index out of bounds.
+            if (dummyTerm < 0) {
```
Comment (munjoonteo) [inline api/src/getStartDate.ts]: This if statement can be combined with the initialisation of `dummyTerm` using the max function. Then the comments can be simplified into a single comment too.

File: api/src/getStartDate.ts [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+                dummyTerm = 0; // Sanity check to make sure the array does not break
+                // Only time this will happen is when the next year's summer starts over.
+            }
+            const oldTermStartDate =
```
Comment (munjoonteo) [inline api/src/getStartDate.ts]: This could be called `currentTermStartDate`. But with a function you can just `res.send()` directly without declaring an intermediate variable.

File: api/src/getStartDate.ts [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+ * Get the latest term start date from data.json
+ */
+const getStartDateNotangles = (req: express.Request, res: express.Response) => {
+    try {
+        // Get the latest term and check the first class of the term
+        // The assumption is that the starting date is the start date of the first class
+        // of the term.
+        const termStartDate =
+            data.timetableData[term()][FIRST_COURSE]["classes"][0]["termDates"]["start"];
+        res.send(termStartDate);
```
Comment (munjoonteo) [inline api/src/getStartDate.ts]: Use a function here to simplify things and avoid declaring one-use variables.

File: api/src/getStartDate.ts [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+    }
+};
+
+/**
+ * Get the term which has data from data.json, this term
+ * is the latest term with data. This is exclusively for Notangles, so
+ * Freerooms please do not use it!
+ * @returns the latest term number
+ */
+const term = () => {
```
Comment (munjoonteo) [inline api/src/getStartDate.ts]: This isn't a great name for a function. Perhaps `getLatestTermNumber`?

File: api/src/getStartDate.ts [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+        res.send(term());
+    } catch (e) {
+        res.status(400).send("Error");
+    }
+};
+
+/**
+ * Get the term which has data from data.json, this term
+ * is the latest term with data. This is exclusively for Notangles, so
+ * Freerooms please do not use it!
```
Comment (munjoonteo) [inline api/src/getStartDate.ts]: This comment is kind of misleading because it's used in `getStartDateFreerooms`

File: api/src/getStartDate.ts [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+ * @returns the latest term number
+ */
+const term = () => {
+    try {
+        let term: string = "";
+        for (let termId of termArray) {
+            let timetableData = data.timetableData;
+            // Check if the termId exists in the timetableData and the fields are actually
+            // present. If they are, return the term as it is.
+            if (timetableData.hasOwnProperty(termId)) {
```
Comment (munjoonteo) [inline api/src/getStartDate.ts]: I think if you reversed the terms in the `termArray` (i.e. start from T3 and go backwards) you can have an early return here and simplify your code. The `term` variable would also be unnecessary I think. Let me know if you can't figure out what I mean. I don't think the reversal would affect the other places where `termArray` is used.

File: api/src/getStartDate.ts [hunk 21]
```diff
@@ -0,0 +1,144 @@
...
+                if (!isClassFound(termId, "ECON1101")) {
+                    break;
+                }
+                term = termId;
+            }
+        }
+
+        return term;
+    } catch (e) {
+        console.log("There was an error getting the most updated term!");
```
Comment (munjoonteo) [inline api/src/getStartDate.ts]: Should this error be re-thrown so a 500 can be returned to the user?

File: api/src/getStartDate.ts [hunk 22]
```diff
@@ -0,0 +1,144 @@
...
+    }
+};
+
+const isClassFound = (termId: string, courseCode: string) => {
+    let timetableData = data.timetableData;
```
Comment (munjoonteo) [inline api/src/getStartDate.ts]: This logic is too long-winded and can be replaced by one of the array functions.

File: api/src/load-data.ts [hunk 23]
```diff
@@ -1,5 +1,5 @@
 import fs from "fs";
-export const dataLocation = "/data/data.json";
+export const dataLocation = "data/data.json";
```
Comment (munjoonteo) [inline api/src/load-data.ts]: Always remember to change this back if not deployment will break!!!!

File: api/src/server.ts [hunk 24]
```diff
@@ -28,6 +29,11 @@ app.use(express.json({ limit: "10mb" }));
 app.get("/api/terms/:termId/courses/:courseId", notangles.getCourse);
 app.get("/api/terms/:termId/courses", notangles.getCourseList);
 app.get("/api/terms/:termId/freerooms", freerooms.getFreeroomsData);
+app.get("/api/startdate/freerooms", getStartDateFreerooms);
+
+app.get("/api/startdate/notangles", getStartDateNotangles);
```
Comment (munjoonteo) [inline api/src/server.ts]: Can you follow the other endpoints and have the Notangles/Freerooms specific functions in the Notangles/Freerooms file? Then maybe add another folder (e.g. `helpers` or something) for files containing helper/shared functions like `getStartDate.ts`.  This is so we can keep the functionality between teams separate and easier to maintain by each team.

Also I think it would be more appropriate to have the `freerooms` and `notangles` paths be a URL parameter instead. Then, have some kind of controller call the appropriate function depending on the value of that parameter.

**Review by munjoonteo — CHANGES_REQUESTED** [review]

File: api/src/getStartDate.ts (hunk 3 above)
Comment (Rayahhhmed) [reply api/src/getStartDate.ts]: Forgot comma too ;-;

File: api/src/getStartDate.ts (hunk 10 above)
Comment (Rayahhhmed) [reply api/src/getStartDate.ts]: matched[0] has the entire regex

File: api/src/getStartDate.ts (hunk 12 above)
Comment (Rayahhhmed) [reply api/src/getStartDate.ts]: Too much 1531 and python

File: api/src/getStartDate.ts (hunk 16 above)
Comment (Rayahhhmed) [reply api/src/getStartDate.ts]: Just did the latter and merged em into one line.

File: api/src/load-data.ts (hunk 23 above)
Comment (Rayahhhmed) [reply api/src/load-data.ts]: SOZ I THOUGHT I CHANGED IT ARGH, will keep a fat reminder for it!

File: api/src/server.ts (hunk 24 above)
Comment (Rayahhhmed) [reply api/src/server.ts]: Will finish this later tonight!

Comment (Rayahhhmed) [reply api/src/server.ts]: I have added this functionality, would love feedback on it!

File: api/src/load-data.ts (hunk 23 above)
Comment (Rayahhhmed) [reply api/src/load-data.ts]: Made sure this was done!

**Review by munjoonteo — APPROVED** [review]

**Review by lhjt — APPROVED** [review]
